File: npm/printobserver-sdk/integration/world.ts
/**
 * The one supervisor all three clients' printer-integration journeys drive.
 *
 * It is stood up by this repository's own tool rather than by each client in
 * its own language: a world built three times would be three worlds, and what
 * the three journeys are for is that the same nine steps against the same
 * supervisor come out the same in all three.
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";

/** Where the supervisor is, and what a journey acts on. */
export interface Supervisor {
  server: string;
  /** The credential it serves under, as the client configuration it wrote carries it. */
  credential: string;
  print_id: string;
  image_id: string;
  event_id: string;
  file_name: string;
}

/** The repository this package is in. */
const REPO_ROOT = resolve(dirname(new URL(import.meta.url).pathname), "../../..");

/**
 * Only the real-server setup hook's budget. Three warmed Linux runs measured
 * 2660, 3016 and 2737 ms, and a macOS runner overran ten seconds standing the
 * same world up, so what bounds it is what the world itself may do rather than
 * a warmed measurement. Both clients read the same committed budget because
 * the first time the tier is driven includes a release build of the program,
 * and this is the same world. Test bodies and cleanup keep their own timeouts.
 */
const startupTimeoutSeconds = Number(
  readFileSync(`${REPO_ROOT}/tools/release-artifacts/world-startup-timeout-seconds`, "utf8").trim(),
);
if (!Number.isSafeInteger(startupTimeoutSeconds) || startupTimeoutSeconds <= 0) {
  throw new Error("the shared world startup timeout must be a positive whole number of seconds");
}
export const SETUP_TIMEOUT_MS = startupTimeoutSeconds * 1000;

/** The packages this repository's own tools live in, read from the justfile. */
function pythonPath(): string {
  for (const line of readFileSync(`${REPO_ROOT}/justfile`, "utf8").split("\n")) {
    if (line.startsWith("export PYTHONPATH :=")) {
      return line.split(":=")[1]?.trim().replaceAll('"', "") ?? "";
    }
  }
  throw new Error("the justfile exports no PYTHONPATH, and this tier's world lives on it");
}

/** A supervisor held up for as long as this is alive. */
export class Standing {
  private readonly holding: ReturnType<typeof Bun.spawn>;
  readonly at: Supervisor;

  private constructor(holding: ReturnType<typeof Bun.spawn>, at: Supervisor) {
    this.holding = holding;
    this.at = at;
  }

  /** Bring one up under `into`, over the scripted OctoPrint. */
  static async standing(into: string): Promise<Standing> {
    const holding = Bun.spawn(
      [
        "uv",
        "run",
        "-q",
        "python",
        "-m",
        "release_artifacts",
        "world",
        "--octoprint",
        "--into",
        into,
      ],
      {
        cwd: REPO_ROOT,
        env: { ...process.env, PYTHONPATH: pythonPath() },
        stdin: "pipe",
        stdout: "pipe",
        stderr: "pipe",
      },
    );

    const said = await firstLine(holding.stdout);
    if (said.trim() === "") {
      const why = await new Response(holding.stderr).text();
      holding.kill();
      throw new Error(
        "the world did not come up. Run `just octoprint-up` first; this tier drives a " +
          `real OctoPrint and has no fixture to fall back to.\n${why}`,
      );
    }
    return new Standing(holding, JSON.parse(said) as Supervisor);
  }

  /** Close the input the world waits on, which is how it is asked to stop. */
  async stop(): Promise<void> {
    const input = this.holding.stdin;
    if (input !== undefined && typeof input !== "number") {
      input.end();
    }
    await this.holding.exited;
  }
}

/** The first line one stream carries. */
async function firstLine(stream: ReadableStream<Uint8Array>): Promise<string> {
  const reader = stream.getReader();
  const decoder = new TextDecoder();
  let said = "";
  while (!said.includes("\n")) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }
    said += decoder.decode(value, { stream: true });
  }
  reader.releaseLock();
  return said.split("\n")[0] ?? "";
}

File: python/printobserver-sdk/integration/world.py
"""The one supervisor all three clients' printer-integration journeys drive.

It is stood up by this repository's own tool rather than by each client in its
own language: a world built three times would be three worlds, and what the
three journeys are for is that the same nine steps against the same supervisor
come out the same in all three.
"""

from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass
from pathlib import Path
from types import TracebackType

from repo_checks.shell import start

REPO_ROOT = Path(__file__).resolve().parents[3]

#: How long the world is given to come up, which includes a release build of
#: the program it runs the first time this tier is driven.
STARTUP_TIMEOUT_SECONDS = 2400


@dataclass(frozen=True, slots=True)
class Supervisor:
    """Where the supervisor is, and what a journey acts on."""

    server: str
    print_id: str
    image_id: str
    file_name: str


def _pythonpath() -> str:
    """The packages this repository's own tools live in, read from the justfile.

    Raises:
        AssertionError: If the justfile exports none, which is a tree these
            tools are not reachable in.
    """
    for line in (REPO_ROOT / "justfile").read_text(encoding="utf-8").splitlines():
        if line.startswith("export PYTHONPATH :="):
            return line.partition(":=")[2].strip().strip('"')
    message = "the justfile exports no PYTHONPATH, and this tier's world lives on it"
    raise AssertionError(message)


class Standing:
    """A supervisor held up for as long as this is entered."""

    def __init__(self, into: Path) -> None:
        """Bring one up under `into`, over the scripted `OctoPrint`."""
        self.into = into
        self._holding: subprocess.Popen[str] | None = None

    def __enter__(self) -> Supervisor:
        """Start it, and say where it is.

        Returns:
            Where the supervisor answers and what a journey acts on.

        Raises:
            AssertionError: If it did not come up, naming the recipe that
                brings the printer environment up: a tier that quietly passed
                against no printer would prove nothing.
        """
        import os

        self._holding = start(
            [
                "uv",
                "run",
                "-q",
                "python",
                "-m",
                "release_artifacts",
                "world",
                "--octoprint",
                "--into",
                str(self.into),
            ],
            cwd=REPO_ROOT,
            env={**os.environ, "PYTHONPATH": _pythonpath()},
        )
        said = self._holding.stdout.readline() if self._holding.stdout else ""
        if not said.strip():
            why = self._holding.stderr.read() if self._holding.stderr else ""
            self._holding.kill()
            message = (
                "the world did not come up. Run `just octoprint-up` first; this tier "
                f"drives a real OctoPrint and has no fixture to fall back to.\n{why}"
            )
            raise AssertionError(message)
        described = json.loads(said)
        return Supervisor(
            server=described["server"],
            print_id=described["print_id"],
            image_id=described["image_id"],
            file_name=described["file_name"],
        )

    def __exit__(
        self,
        kind: type[BaseException] | None,
        value: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        """Close the input the world waits on, which is how it is asked to stop."""
        if self._holding is None:
            return
        if self._holding.stdin is not None:
            self._holding.stdin.close()
        try:
            self._holding.wait(timeout=120)
        except subprocess.TimeoutExpired:
            self._holding.kill()
            self._holding.wait(timeout=120)
        self._holding = None
